File: demos/android/src/main/java/com/antgroup/antv/f2/samples/ChartListActivity.java
package com.antgroup.antv.f2.samples;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.antgroup.antv.f2.F2CanvasView;
import com.antgroup.antv.f2.F2Log;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

public class ChartListActivity extends AppCompatActivity {

    private RecyclerView listView;

    private List<WeakReference<F2CanvasView>> canvasViews = new ArrayList<>();


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        setContentView(R.layout.activity_main);

        listView = findViewById(R.id.listView);

        listView.setLayoutManager(new LinearLayoutManager(this));
        listView.setAdapter(new CanvasViewAdapter(MainActivity.getChartModels(), canvasViews));
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        F2Log.i("ChartListActivity", "canvasViews Size: " + canvasViews.size());
        for (WeakReference<F2CanvasView> canvasView : canvasViews) {
            if (canvasView.get() != null) {
                canvasView.get().destroy();
            }
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            finish();
        }
        return super.onOptionsItemSelected(item);
    }

    static class CanvasViewHolder extends RecyclerView.ViewHolder {

        TextView chartName;
        F2CanvasView canvasView;

        public CanvasViewHolder(@NonNull View itemView, List<WeakReference<F2CanvasView>> canvasViews) {
            super(itemView);
            chartName = itemView.findViewById(R.id.chart_name);
            canvasView = itemView.findViewById(R.id.canvasView);
            canvasView.initCanvasContext(new F2CanvasView.ConfigBuilder()
                    .setOption("canvasBizId", "ChartListActivity")
                    .setOption("appId", "1000")
                    .build());
            canvasViews.add(new WeakReference<>(canvasView));
        }
    }


    static class CanvasViewAdapter extends RecyclerView.Adapter<CanvasViewHolder> {
        private List<ChartModel> mChartModels;
        private List<WeakReference<F2CanvasView>> mCanvasViews;

        CanvasViewAdapter(List<ChartModel> chartModels, List<WeakReference<F2CanvasView>> canvasViews) {
            mChartModels = chartModels;
            mCanvasViews = canvasViews;
        }

        @NonNull
        @Override
        public CanvasViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            View itemView = LayoutInflater.from(parent.getContext()).inflate(R.layout.layout_chart_canvasview, parent, false);
            return new CanvasViewHolder(itemView, mCanvasViews);
        }

        @Override
        public void onBindViewHolder(@NonNull CanvasViewHolder holder, int position) {
            final ChartModel chartModel = mChartModels.get(position);
            holder.chartName.setText(chartModel.title);

            try {
                F2CanvasView.Adapter adapter = chartModel.adapterClass.newInstance();
                holder.canvasView.setAdapter(adapter);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }

        @Override
        public int getItemCount() {
            return mChartModels.size();
        }
    }
}

File: demos/android/demos-kotlin/src/main/java/com/antgroup/antv/f2/kotlinsamples/ChartListActivity.kt
package com.antgroup.antv.f2.kotlinsamples

import android.os.Bundle
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.ViewHolder
import com.antgroup.antv.f2.F2CanvasView
import com.antgroup.antv.f2.F2Log
import com.antgroup.antv.f2.kotlinsamples.MainActivity.Companion.chartModels
import java.lang.ref.WeakReference
import java.util.*

class ChartListActivity : AppCompatActivity() {
    private var listView: RecyclerView? = null
    private val canvasViews: MutableList<WeakReference<F2CanvasView?>> = ArrayList()
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        setContentView(R.layout.activity_main)
        listView = findViewById(R.id.listView) as RecyclerView
        listView!!.setLayoutManager(LinearLayoutManager(this))
        listView!!.setAdapter(CanvasViewAdapter(chartModels, canvasViews))
    }

    override fun onDestroy() {
        super.onDestroy()
        F2Log.get().i("ChartListActivity", "canvasViews Size: " + canvasViews.size)
        for (canvasView in canvasViews) {
            if (canvasView.get() != null) {
                canvasView.get()!!.destroy()
            }
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            finish()
        }
        return super.onOptionsItemSelected(item)
    }

    internal class CanvasViewHolder(
        itemView: View,
        canvasViews: MutableList<WeakReference<F2CanvasView?>>
    ) : ViewHolder(itemView) {
        var chartName: TextView
        var canvasView: F2CanvasView

        init {
            chartName = itemView.findViewById(R.id.chart_name)
            canvasView = itemView.findViewById(R.id.canvasView)
            canvasView.setCanvasBizId("ChartListActivity")
            canvasViews.add(WeakReference(canvasView))
        }
    }

    internal class CanvasViewAdapter(
        private val mChartModels: List<ChartModel>,
        private val mCanvasViews: MutableList<WeakReference<F2CanvasView?>>
    ) : RecyclerView.Adapter<CanvasViewHolder>() {
        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CanvasViewHolder {
            val itemView = LayoutInflater.from(parent.context)
                .inflate(R.layout.layout_chart_canvasview, parent, false)
            return CanvasViewHolder(itemView, mCanvasViews)
        }

        override fun onBindViewHolder(holder: CanvasViewHolder, position: Int) {
            val chartModel = mChartModels[position]
            holder.chartName.text = chartModel.title
            try {
                val adapter = chartModel.adapterClass.newInstance()!!
                holder.canvasView.setAdapter(adapter)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }

        override fun getItemCount(): Int {
            return mChartModels.size
        }
    }
}